### PR TITLE
feat: upgrade ANN energy predictor training pipeline, notebook, and docs

### LINKED ISSUE #2146: 💡[FEATURE]: Populate train.py pipeline, enable model persistence, and upgrade documentation
## 📌 Overview
Currently, the core project files in `ANN_Energy_Consumption_Predictor` are incomplete or lack production-grade features. This issue proposes upgrading the existing files (`train.py`, `requirements.txt`, `README.md`, and `energy_predictor.ipynb`) without adding any new files to the repository.

---

## 🔍 Current Issues & Limitations

1. **Empty `train.py`**: The `train.py` file is currently empty (`0 bytes`), forcing all training to be done manually inside the Jupyter Notebook.
2. **Missing Model & Scaler Persistence**: Trained model weights and feature scalers are not saved to disk, making model inference on new data impossible without retraining.
3. **Data Subsampling in Notebook**: The notebook samples only 5,000 rows out of 19,735 (~75% data loss), reducing model accuracy and generalization capability.
4. **Minimal Documentation & Unpinned Dependencies**:
   - `requirements.txt` lacks version bounds.
   - `README.md` is minimal and lacks CLI setup and execution instructions.

---

### 1. `train.py`
- Build an end-to-end training script utilizing the **full dataset** (19,735 rows).
- Add feature engineering: time components (`hour`, `day`, `month`, `dayofweek`), cyclical encodings (`sin_hour`, `cos_hour`), and temperature/humidity interaction terms.
- Train an ANN regression model with TensorFlow/Keras using `EarlyStopping` and `ReduceLROnPlateau`.
- Save model weights (`results/energy_model.keras`) and scaler (`results/scaler.pkl`) inside `results/`.
- Add a `--predict` CLI flag directly inside `train.py` to allow running predictions on sample data without needing a separate prediction script.

### 2. `requirements.txt`
- Pin version bounds for core packages: `pandas`, `numpy`, `matplotlib`, `scikit-learn`, `tensorflow`, `jupyter`, `joblib`.

### 3. `README.md`
- Provide complete project setup instructions.
- Add CLI execution guide for training and prediction using `train.py`.
- Document dataset features and model evaluation metrics ($R^2$, MAE, RMSE).

### 4. `energy_predictor.ipynb`
- Update notebook to train on the full dataset, eliminate convergence warnings, and improve visualization plots.

---

## ✅ Expected Outcomes
- Fully functional CLI training & prediction via `train.py`.
- Model artifact persistence for future inference.
- Improved prediction accuracy using full dataset training.

### PR DESCRIPTION
**Pull request by Tanishk-Rastogi**
## Summary

Closes #2146

This PR enhances the existing `ANN_Energy_Consumption_Predictor` by upgrading the core project files (`train.py`, `energy_predictor.ipynb`, `requirements.txt`, and `README.md`) without introducing any new source files.

The updates include full-dataset training, improved feature engineering, model artifact persistence, CLI-based inference, enhanced notebook visualizations, and expanded project documentation.

---

### `train.py`

* Trains on the full dataset (19,735 samples).
* Added feature engineering:

  * Time-based features (`hour`, `day`, `month`, `dayofweek`)
  * Cyclical encodings (`sin_hour`, `cos_hour`)
  * Indoor environmental averages
  * Living room interaction feature (`T1 × RH_1`)
* Replaced the previous training pipeline with a TensorFlow/Keras Sequential regression model.
* Added `EarlyStopping` and `ReduceLROnPlateau` callbacks for improved convergence.
* Automatically saves:

  * `results/energy_model.keras`
  * `results/scaler.pkl`
  * `results/metrics.json`
  * Evaluation plots
* Added CLI inference through the `--predict` flag for sample data and custom CSV inputs.

### `energy_predictor.ipynb`

* Removed dataset subsampling and trained on the complete dataset.
* Replaced `MLPRegressor` with a Keras Sequential model.
* Added:

  * Training and validation loss curves
  * Actual vs. predicted scatter plot
  * Time-series comparison plot

### `requirements.txt`

* Added explicit minimum versions for:

  * pandas
  * numpy
  * matplotlib
  * scikit-learn
  * tensorflow
  * jupyter
  * joblib

### `README.md`

* Expanded setup instructions.
* Added CLI training and prediction examples.
* Documented engineered features.
* Included evaluation metrics (R², MAE, RMSE, and MAPE).

---

### Training

```bash
python train.py --epochs 50 --batch_size 64
```

Verified that the following artifacts are generated successfully:

* `energy_model.keras`
* `scaler.pkl`
* `metrics.json`
* Training history plot
* Prediction scatter plot

### Inference

```bash
python train.py --predict
```

Verified successful inference using the saved model and scaler on sample input data.

---

## Results

* Successfully trains on the complete dataset.
* Training converges without warnings.
* Model artifacts are saved correctly and can be reused for inference.
* CLI prediction works as expected using the persisted model.

### CODE REVIEW

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
remove all the pngs and make pdf and explain the pngs 

@Tanishk-Rastogi

**Review by Sibam-Paul — APPROVED** [review]
